packages/react/context-menu/src/ContextMenu.tsx

```diff
@@ -20,12 +22,42 @@ type ContextMenuContextValue = {
   onOpenChange(open: boolean): void;
 };
 
+const SubmenuContext = React.createContext<boolean | undefined>(undefined);
+const InsideContentContext = React.createContext<boolean>(false);
 const [ContextMenuProvider, useContextMenuContext] = createContext<ContextMenuContextValue>(
   CONTEXT_MENU_NAME
 );
 
-const ContextMenu: React.FC<{ onOpenChange?(open: boolean): void }> = (props) => {
-  const { children, onOpenChange } = props;
+type ContextMenuOwnProps = ContextMenuRootOwnProps & ContextMenuSubOwnProps;
+
+const ContextMenu: React.FC<ContextMenuOwnProps> = (props) => {
+  const parentSubmenuContext = React.useContext(SubmenuContext);
+  const isInsideContent = React.useContext(InsideContentContext);
+  const isRootMenu = parentSubmenuContext === undefined;
+
+  return (
+    <SubmenuContext.Provider value={!isRootMenu && isInsideContent}>
+      <ContextMenuImpl {...props} />
+    </SubmenuContext.Provider>
+  );
+};
+
+const ContextMenuImpl: React.FC<ContextMenuOwnProps> = (props) => {
+  const isSubmenu = React.useContext(SubmenuContext);
+  const ContextMenu = isSubmenu ? ContextMenuSub : ContextMenuRoot;
+
+  return <ContextMenu {...props} />;
```

Comment:
It seems we tend to prefer to do the ternary inline to see both components being used.

Reply:
Haha yeah we've done a bit of both on this one. For a bit of backstory...

I think we used to do it this way but we got tripped up a few times trying to read the code. In `Slider` for example we have `const SliderOrientation = condition ? SliderVertical : SliderHorizontal;` and I always read the render and see `<SliderOrientation />` and then try to find a `SliderOrientation` part that doesn't exist 😅 

We started to favour slight duplication to ease this confusion.